File: C++/queue_circular.cpp
#include <stdio.h>
#include <stdlib.h>
#define max 100
int queue[max], n, x;
int front = -1;
int rear = -1;
void enqueue()
{
    if (front == -1 && rear == -1)
    {
        front = 0;
        rear = 0;
        printf("Enter the element : ");
        scanf("%d", &x);
        queue[rear] = x;
    }
    else if ((rear + 1) % n == front)
        printf("Queue is full !!!");
    else
    {
        rear = (rear + 1) % n;
        printf("Enter the element : ");
        scanf("%d", &x);
        queue[rear] = x;
    }
}
void dequeue()
{
    int item;
    if (front == -1 && rear == -1)
        printf("Queue is empty !!!");
    else if (front == rear)
    {
        item = front;
        front = -1;
        rear = -1;
    }
    else
    {
        item = front;
        front = (front + 1) % n;
    }
}
void peek()
{
    if (front == -1 && rear == -1)
        printf("Queue is empty !!!");
    else
    {
        printf("The front element is %d", queue[front]);
    }
}
void display()
{
    int i = front;
    if (front == -1 && rear == -1)
        printf("Queue is empty !!!");
    else
    {
        printf("Queue : ");
        while (i != rear)
        {
            printf("%d ", queue[i]);
            i = (i + 1) % n;
        }
        printf("%d", queue[rear]);
    }
}
int main()
{
    int op;
    printf("Enter the size of the queue : ");
    scanf("%d", &n);
    while (1)
    {
        printf("\n1. Insert an element\n2. Delete an element\n3. Peek\n4. Display the elements\n5. Exit\n\nEnter your choice : ");
        scanf("%d", &op);
        switch (op)
        {
        case 1:
            enqueue();
            break;
        case 2:
            dequeue();
            break;
        case 3:
            peek();
            break;
        case 4:
            display();
            break;
        case 5:
            exit(0);

        default:
            printf("Wrong choice !!! Try again\n");
            break;
        }
    }
}

File: Python/queue_circular.py
front, rear = -1, -1
n = int(input("Enter the size of the queue : "))
queue_c = [None for _ in range(n)]


def enqueue():
    global front, rear
    if front == -1 and rear == -1:
        front, rear = 0, 0
        x = int(input("Enter the element : "))
        queue_c[rear] = x
    elif (rear + 1) % n == front:
        print("Queue is full")
    else:
        rear = (rear + 1) % n
        x = int(input("Enter the element : "))
        queue_c[rear] = x


def deque():
    global front, rear
    if front == -1 and rear == -1:
        print("Queue is empty")
    elif front == rear:
        front = -1
        rear = -1
    else:
        front = (front + 1) % n
        print("First element is removed from the queue")


def display():
    i = front
    if front == -1 and rear == -1:
        print("Queue is empty")
    else:
        while i != rear:
            print(queue_c[i], end=" ")
            i = (i + 1) % n
        print(queue_c[rear])


print("\n1. Insert an element\n2. Delete an element\n3. Display the elements\n4. Exit\n")
while True:
    op = int(input("Enter your choice : "))
    if op == 1:
        enqueue()
    elif op == 2:
        deque()
    elif op == 3:
        display()
    elif op == 4:
        quit()
    else:
        print("Wrong choice !!! Try again")
